data-science-onramp/data-cleaning/clean.py

```diff
@@ -123,69 +123,61 @@ def compute_end_udf(duration, start, end):
+        print(f'{TABLE} does not exist.')
+        raise 
 
-    # Single-parameter column transformations
+    # Single-parameter udfs
     udfs = {
-        'start_station_name': (station_name_udf, StringType()),
-        'end_station_name': (station_name_udf, StringType()),
-        'tripduration': (trip_duration_udf, IntegerType()),
-        'usertype': (user_type_udf, StringType()),
-        'gender': (gender_udf, StringType()),
-        'start_station_latitude': (angle_udf, FloatType()),
-        'start_station_longitude': (angle_udf, FloatType()),
-        'end_station_latitude': (angle_udf, FloatType()),
-        'end_station_longitude': (angle_udf, FloatType())
+        'start_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'end_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'tripduration': UserDefinedFunction(trip_duration_udf, IntegerType()),
+        'usertype': UserDefinedFunction(user_type_udf, StringType()),
+        'gender': UserDefinedFunction(gender_udf, StringType()),
+        'start_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'start_station_longitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_longitude': UserDefinedFunction(angle_udf, FloatType())
     }
 
     for name, udf in udfs.items():
-        df = df.withColumn(name, UserDefinedFunction(*udf)(name))
+        df = df.withColumn(name, udf(name))
 
-    # Multiple-parameter column transformations
+    # Multi-parameter udfs
     multi_udfs = {
         'tripduration': {
-            'udf': (compute_duration_udf, IntegerType()),
+            'udf': UserDefinedFunction(compute_duration_udf, IntegerType()),
             'params': ('tripduration', 'starttime', 'stoptime')
         },
         'starttime': {
-            'udf': (compute_start_udf, StringType()),
+            'udf': UserDefinedFunction(compute_start_udf, StringType()),
             'params': ('tripduration', 'starttime', 'stoptime')
         },
         'stoptime': {
-            'udf': (compute_end_udf, StringType()),
+            'udf': UserDefinedFunction(compute_end_udf, StringType()),
             'params': ('tripduration', 'starttime', 'stoptime')
         }
     }
 
     for name, obj in multi_udfs.items():
-        df = df.withColumn(name, UserDefinedFunction(*obj['udf'])(*obj['params']))
+        df = df.withColumn(name, obj['udf'](*obj['params']))
 
-    if upload:
-        upload_to_gcs(df)
-    
+    # Display sample of 100 rows
     df.sample(False, 0.001).show(n=100)
 
-if __name__ == '__main__':
-    main()
+    # Write results to GCS
+    if '--test' in sys.argv:
+        print('Data will not be uploaded to GCS')
+    else:
+        path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+        df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path, mode='overwrite')
```

Comment:
@TusharK54 A couple things here:

1. Can you add `.coalesce(1)` such that you have `df.coalesce(1).write...`? This will condense your output into a single file. You may also want to split this up over multiple lines to avoid the line-length issue:
```
(
    df
    .coalesce(1)
    .write
    ...
)

2. Can you please add logic such that this first writes out to a tmp file and then is copied over to a more conventional location? Feel free to reuse the logic [here](https://github.com/GoogleCloudDataproc/cloud-dataproc/blob/master/codelabs/spark-bigquery/backfill.py#L97)


Can you add logic that does something similar to this? `https://github.com/GoogleCloudDataproc/cloud-dataproc/blob/master/codelabs/spark-bigquery/backfill.py#L97`

That is, save the data to a temporary GCS bucket, and then move it to this bucket but change the naming to something more conventional.

Reply:
Update: Can you and @vuppalli work together to decide which approach makes sense (my suggestion, as is or something entirely different)? Since this will inevitably impact both of your projects, it'l be good for you to be on the same page.

Reply:
I think that adding `.coalesce(1)` might make the script run longer. Since the script already takes a while to complete, and if adding this will make it longer, this might not be a good idea. What do you think? @TusharK54

Reply:
@bradmiro to address your comments:
1. I just tested it and it seems @vuppalli is right. Using coalesce results in the job taking twice as long as without. Is that a deal breaker?
2. I will go ahead and do that. But could you let me know why we do it like that?